connection_unix.go

```diff
@@ -252,6 +252,23 @@ func (c *conn) Wake() error {
 	})
 }
 
+func (c *conn) AsyncExecute(cb func(Conn) ([]byte, Action)) error {
```

Comment:
Sorry for bringing it up so late, but after a second thought, I find `AsyncExecute` may not be quite right here, since users are allowed to make their business logic be executed within the I/O event-loop, which may occupy the CPU for a long time, starving the subsequent I/O events.

Reply:
I didn't think it through before, but now I've realized it and we should start a more profound discussion about it.

Reply:
That's a fair point! I have been some thoughts about it too. But on another hand, I don't see here much difference between `React` which can use locks or something like external queries, and `AsyncExecute` which have the same limitations but can be initiated externally. Both of them should have the same recommendations about its usage (do not use locks or blocking calls, offload them to pools or goroutines).
I understand your concern about the potential misuse of that method. I'll think about ways how to reduce the effect of the wrong usage of that methods. But still, I think that `React` has the same limitations and troubles as` AsyncExecute` has.

Reply:
Blocking jobs shouldn't be executed inside `React()`, those jobs are supposed to go to separate goroutines and call AsyncWrite() after it's done.
If we add `AsyncExecute` method, users tend to put all their business logic into that callback, which results in the blocking IO event-loops.

Reply:
Maybe interface reduced to `UpdateCtx(ctx interface{}) (new interface{})` can be less dangerous? It will still provide ability to consistent update ctx but prevent users from overuse it. It still possible of far, but at least not that easy as in current proposal.

Reply:
Updating the context without mutex is your initial  requirement, but I wonder that mutex was really giving you a hard time here?

Reply:
Mutex itself is not that hard as a requirement to store and maintain another map with concurrent access.

A little more details about my needs:
I have long-lived WebSocket connections. I need to write into another system uptime and the number of messages that each connection received. Periodically and when the connection stops.

I definitely can do that if I introduce a map that will be stored separately. When a connection opened it will acquire a mutex and write itself into a map. When the connection stops — it will acquire mutex again and delete itself. Periodically I will iterate over the map (with mutex locked) and push uptime into the needed system (mutex can be unlocked here). That will work. But it introduces at least one disadvantage here — it reduces the efficiency of the multiple reactors and raises contention that can make mutex slow.

For example:
I have 12 reactors. So when I receive a connection I should acquire a mutex. But if another connection already acquires it I should wait. And if I have a pretty large number of incoming connections they all should wait for each other. So only one conn can be accepted, even if I use multiple reactors and reuse port which should allow me to accept at least 12 conns at a time.
And sometimes, when I will iterate over the map to send registered uptimes, all accepts will be paused.
As a way to reduce contention between reactors, I can introduce shards that will require 12 mutexes and 12 maps. But to do that I should use reflection to get loop index from a conn (I already do that to be clear, so it would be great to have a way to have an API to get it), otherwise, if I do not aware of the loop index I still have a risk to raise contention if some of the connections bound to different reactors will get the same shard.

Even that explanation sounds a little bit complicated I think :)
If I have a way to update context I should do nothing more than trigger update when I want to write something to a conn and trigger send to an external system when the conn closed. Periodical updates can be done in the same way (I can initiate send from UpdateCtx func, that solution a little bit dirty but do its work). 
With `AsyncExecute` this can be done even simpler, but I understand that `AsyncExecute` introduces a way to misuse API and slow down an entire server.

Reply:
If you prefer to think yet about implementation that's not a problem, I'll use own fork for myself and you have a time to decide how to do that with better API. My proposal covers my needs mostly, it may really require safer way to do that.

Reply:
I really appreciate your understanding here, ligser, thanks~

Reply:
I'll spend some time later to think about this requirement and a suitable solution for it.